### PR TITLE
Handle param null check with else clause

### PR DESCRIPTION
**Pull request by RikkiGibson**
Related to #36024

Gives a correct fix in a scenario like the following:
MicrosoftTeams-image (3)

cc @CyrusNajmabadi

### CODE REVIEW

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs [hunk 1]
```diff
@@ -60,6 +60,33 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                            }
+                            for (var i = indexOfIfStatement + 1; i < statements.Count; i++)
+                            {
+                                newStatements.Add(statements[i]);
+                            }
+
+                            parent.Statements.ReplaceRange(ifStatementWithElseBlock, newStatements);
+                            editor.ReplaceNode(parent, parent.WithStatements(SyntaxFactory.List(newStatements)));
+                            break;
+                        }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:85]: my general preference is once switch clauses get large, they should be broken out into helper methods.

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs [hunk 2]
```diff
@@ -60,6 +60,33 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
                         var parameterReferenceSyntax = nullCoalescing.Left;
                         editor.ReplaceNode(nullCoalescing, parameterReferenceSyntax.WithAppendedTrailingTrivia(SyntaxFactory.ElasticMarker));
                         break;
+                    case IfStatementSyntax { Else.Statement: BlockSyntax { Statements: var statementsWithinElse } } ifStatementWithElseBlock:
+                        {
+                            var parent = (BlockSyntax)ifStatementWithElseBlock.Parent!;
+                            var statements = parent.Statements;
+                            var indexOfIfStatement = statements.IndexOf(ifStatementWithElseBlock);
+                            using var _1 = ArrayBuilder<StatementSyntax>.GetInstance(capacity: statements.Count + statementsWithinElse.Count - 1, out var newStatements);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:68]: could not care less about hte capacity.  would be happy for you to remove that just for code cleanliness.

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs [hunk 3]
```diff
@@ -60,6 +60,33 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
+
+                            for (var i = 0; i < indexOfIfStatement; i++)
+                            {
+                                newStatements.Add(statements[i]);
+                            }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:73]: can this first loop be replaced with some sort of AddRange?

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs [hunk 4]
```diff
@@ -60,6 +60,33 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                            foreach (var statementWithinElse in statementsWithinElse)
+                            {
+                                newStatements.Add(statementWithinElse.WithPrependedLeadingTrivia(SyntaxFactory.ElasticMarker));
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:76]: would be good to have some tests with trivia on these nodes.

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs [hunk 5]
```diff
@@ -60,6 +60,33 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                            }
+                            for (var i = indexOfIfStatement + 1; i < statements.Count; i++)
+                            {
+                                newStatements.Add(statements[i]);
+                            }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:81]: can we do this with an AddRange of some sort?

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs [hunk 6]
```diff
@@ -60,6 +60,33 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
                         var parameterReferenceSyntax = nullCoalescing.Left;
                         editor.ReplaceNode(nullCoalescing, parameterReferenceSyntax.WithAppendedTrailingTrivia(SyntaxFactory.ElasticMarker));
                         break;
+                    case IfStatementSyntax { Else.Statement: BlockSyntax { Statements: var statementsWithinElse } } ifStatementWithElseBlock:
+                        {
+                            var parent = (BlockSyntax)ifStatementWithElseBlock.Parent!;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:65]: how do you know hte parent is a block?  was that tested on the analyzer side?

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs [hunk 7]
```diff
@@ -60,6 +60,33 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                            foreach (var statementWithinElse in statementsWithinElse)
+                            {
+                                newStatements.Add(statementWithinElse.WithPrependedLeadingTrivia(SyntaxFactory.ElasticMarker));
+                            }
+                            for (var i = indexOfIfStatement + 1; i < statements.Count; i++)
+                            {
+                                newStatements.Add(statements[i]);
+                            }
+
+                            parent.Statements.ReplaceRange(ifStatementWithElseBlock, newStatements);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs]: this line appears to have no purpose afaict.

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:85]: (i would extract both 'ifstatement' cases into one case that delegates to a simple helper that then checks the shape of it).

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs (hunk 7 above)
Comment (RikkiGibson) [reply src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs]: That's correct. Removed.

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs (hunk 6 above)
Comment (RikkiGibson) [reply src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:65]: We only search statements directly contained within the method block body.

https://github.com/dotnet/roslyn/blob/c83e59dc080d619868e9bae7c79a45104b26c6d1/src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs#L142

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs]: Simplifying this. I think this also addresses several comments below.

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs]: terrific!

**Comment by RikkiGibson** [comment]
@CyrusNajmabadi I believe I have addressed all your review comments. Please let me know if you have any others.

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs [hunk 8]
```diff
@@ -56,10 +57,17 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
                 switch (node)
                 {
                     case ExpressionSyntax { Parent: BinaryExpressionSyntax(SyntaxKind.CoalesceExpression) nullCoalescing }:
-                        // this.item = item ?? throw new ArgumentNullException(nameof(item));
                         var parameterReferenceSyntax = nullCoalescing.Left;
                         editor.ReplaceNode(nullCoalescing, parameterReferenceSyntax.WithAppendedTrailingTrivia(SyntaxFactory.ElasticMarker));
                         break;
+                    case IfStatementSyntax { Else.Statement: BlockSyntax { Statements: var statementsWithinElse } } ifStatementWithElseBlock:
+                        var parent = (BlockSyntax)ifStatementWithElseBlock.Parent!;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:64]:
```suggestion
                        var parent = (BlockSyntax)ifStatementWithElseBlock.GetRequiredParent();
```

File: src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs [hunk 9]
```diff
@@ -56,10 +57,17 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                        var newStatements = parent.Statements.ReplaceRange(ifStatementWithElseBlock, statementsWithinElse.Select(s => s.WithPrependedLeadingTrivia(SyntaxFactory.ElasticMarker)));
+                        editor.ReplaceNode(parent, parent.WithStatements(newStatements));
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:66]: much nicer! :)

File: src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs [hunk 10]
```diff
@@ -1732,6 +1732,210 @@ public void M(string s!!, int i = 0)
...
+        }
+    }
+}",
+                FixedCode = @"using System;
+public class C
+{
+    public void M(string s1!!)
+    {
+        /* comment1 */
+        s1.ToString() /* comment2 */; // comment3
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:1932]: interesting.  but seems acceptable for now.  we really need an annotation which is "just fixup the indentation, nothing else".

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:1932]: actually, can you file a workitem for that, and link back to this as an example?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs (hunk 10 above)
Comment (RikkiGibson) [reply src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:1932]: Opened #59228

**Comment by RikkiGibson** [comment]
@CyrusNajmabadi is there a need to handle indent depth of new lines within statements of the else clause? (i.e. by adding elastic trivia) in a scenario like the following?

MicrosoftTeams-image (7)

MicrosoftTeams-image (8)

 cc @stephentoub

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi is there a need to handle indent depth of new lines within statements of the else clause? (i.e. by adding elastic trivia) in a scenario like the following?

Yup, that's the same issue i had you open.  This is why we need a general "dedent/indent construct to appropriate level" annotation, which *only* does that work and no other formatting.

**Comment by RikkiGibson** [comment]
#59228

**Pull request merged**